r/src/io.cpp

```diff
@@ -207,7 +209,209 @@ void io___BufferOutputStream__Write(
   StopIfNotOk(stream->Write(RAW(bytes), bytes.size()));
 }
 
-// TransformInputStream::TransformFunc wrapper
+// ------ RConnectionInputStream / RConnectionOutputStream
+
+class RConnectionFileInterface : public virtual arrow::io::FileInterface {
+ public:
+  explicit RConnectionFileInterface(cpp11::sexp connection_sexp)
+      : connection_sexp_(connection_sexp), closed_(false) {
+    check_closed();
+  }
+
+  arrow::Status Close() {
+    if (closed_) {
+      return arrow::Status::OK();
+    }
+
+    auto result = SafeCallIntoR<bool>([&]() {
+      cpp11::package("base")["close"](connection_sexp_);
+      return true;
+    });
+
+    RETURN_NOT_OK(result);
+    closed_ = true;
+    return arrow::Status::OK();
+  }
+
+  arrow::Result<int64_t> Tell() const {
+    if (closed()) {
+      return arrow::Status::IOError("R connection is closed");
+    }
+
+    cpp11::sexp result = cpp11::package("base")["seek"](connection_sexp_);
+    return cpp11::as_cpp<int64_t>(result);
+  }
+
+  bool closed() const { return closed_; }
+
+ protected:
+  cpp11::sexp connection_sexp_;
+
+  // Define the logic here because multiple inheritance makes it difficult
+  // for this base class, the InputStream and the RandomAccessFile
+  // interfaces to co-exist.
+  arrow::Result<int64_t> ReadBase(int64_t nbytes, void* out) {
+    if (closed()) {
+      return arrow::Status::IOError("R connection is closed");
+    }
+
+    return SafeCallIntoR<int64_t>([&] {
```

Comment:
In the category of "probably not worth the effort & complexity but pedantic note for the purists", it might be *slightly* nicer if you overloaded the async versions of the `arrow::io::FileInterface` methods and changed the sync versions to call the async verisons (instead of vice-versa).

The only real gain is that you can avoid spinning up an I/O thread for no reason.  I/O threads are supposed to block on long operations so it isn't really a problem but it is a slight bit of extra overhead that isn't necessary.

![safe-call-into-r-async](https://user-images.githubusercontent.com/1696093/162872351-6e24e9bd-dbf2-4ed1-9d48-c06691d2d944.png)

In fact, now that I draw this picture, I wonder if there might be some way to handle this by having an "R filesystem" whose I/O context's executor was the R main thread :thinking: .  Not something we need to tackle right now.

Reply:
No need for another try.  As I said, this is venturing into somewhat pedantic waters.